### PR TITLE
DOC: Fixes DataFrame.replace documentation

### LINKED ISSUE #55570: DOC: DataFrame.replace() regex contract inconsistent with usage
### Pandas version checks

- [X] I have checked that the issue still exists on the latest versions of the docs on `main` [here](https://pandas.pydata.org/docs/dev/)

### Location of the documentation

https://pandas.pydata.org/docs/dev/reference/api/pandas.DataFrame.replace.html

### Documentation problem

From https://pandas.pydata.org/docs/dev/reference/api/pandas.DataFrame.replace.html I quote:

> **regex : bool or same types as to_replace, default False**
> Whether to interpret _to_replace_ and/or value as regular expressions. If this is `True` then _to_replace_ must be a string. Alternatively, this could be a regular expression or a list, dict, or array of regular expressions in which case _to_replace_ must be `None`.

I want to bring your attention to the first condition:
> If this is `True` then _to_replace_ must be a string.

However, in all the Pandas versions I've tried, this doesn't appear to be the case. The `to_replace` argument need _not_ be a string when supplying `regex=True`, and providing a nested dict in `to_replace` appears to work as expected.

MWE:

```python
import pandas as pd

# Create a sample DataFrame
data = {
    'A': [1, 2, 3, 4, 5],
    'B': [10, 20, 30, 40, 50],
    'C': ['apple', 'banana', 'cherry', 'date', 'elderberry']
}
df = pd.DataFrame(data)

# Display the original DataFrame
print("Original DataFrame:")
print(df)

# Test regex replacements using a nested dictionary
nested_dict = {
    'C': {
        r'^[a-d]': 'fruit_',  # Replace words starting with 'a' to 'd'
        r'[^aeiou]+$': 'berry',  # Replace words not ending in a vowel with 'berry'
    }
}

# Apply regex replacements using the nested dictionary
df.replace(to_replace=nested_dict, regex=True, inplace=True)

# Display the modified DataFrame with regex replacements
print("\nModified DataFrame:")
print(df)
```

### Suggested fix for documentation

Update documentation or API behaviour.

### PR DESCRIPTION
**Pull request by jpgianfaldoni**
- [x] closes #55570 
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/shared_docs.py [hunk 1]
```diff
@@ -570,8 +570,11 @@
         .. deprecated:: 2.1.0
     regex : bool or same types as `to_replace`, default False
         Whether to interpret `to_replace` and/or `value` as regular
-        expressions. If this is ``True`` then `to_replace` *must* be a
-        string. Alternatively, this could be a regular expression or a
+        expressions. If this is ``True``, `value` is not ``None`` and `to_replace` is a string,
```
Comment (mroeschke) [inline pandas/core/shared_docs.py:573]: Can you describe this in the `Example`s section below instead?

Comment (jpgianfaldoni) [reply pandas/core/shared_docs.py:573]: Sure!

Comment (jpgianfaldoni) [reply pandas/core/shared_docs.py]: @mroeschke please let me know if i can do any improvements/corrections!

Comment (mroeschke) [reply pandas/core/shared_docs.py]: Please reference the code checks and pre-commit CI jobs for some formatting issues that may exist here

**Review by jpgianfaldoni — COMMENTED** [review]
pre-commit.ci autofix

File: pandas/core/shared_docs.py [hunk 2]
```diff
@@ -790,6 +789,32 @@
...
+
+    >>> df.replace(to_replace='^[a-g]', value = 'e', regex=True)
+        A  B  C
+    0  0  e  e
+    1  1  e  e
+    2  2  e  h
+    3  3  e  i
+    4  4  e  j
+
+    If `value` is not ``None`` and `to_replace` is a dictionary, the dictionary
```
Comment (mroeschke) [inline pandas/core/shared_docs.py:808]:
```suggestion
    If ``value`` is not ``None`` and `to_replace` is a dictionary, the dictionary
```

**Review by mroeschke — COMMENTED** [review]
Minor comment otherwise LGTM

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jpgianfaldoni

**Pull request merged**